### PR TITLE
Use schema information for all REST operations

### PR DESCRIPTION
**Pull request by lblackstone**
Previously, the provider computed a resource ID during the Create operation, and then used this ID with hardcoded HTTP verbs for Read, Update, and Delete operations. While this works for many resources, some don’t follow this pattern. This change refactors the provider logic to use the endpoints and HTTP verbs specified for each operation by the schema.

---

I computed a [metadata diff](https://gist.github.com/lblackstone/52f66a0e9e9c4e819d999e9f3a0c1f8e) compared to the `v0.15.0` release. There are a bunch of changes to resources with a `IamPolicy` suffix that change `get` to `set` on update methods. The old metadata was not accurate for these cases, and the provider [previously handled them with an override](https://github.com/pulumi/pulumi-google-native/blob/v0.15.0/provider/pkg/provider/provider.go#L645-L647). 

Beyond that, the main changes are to HTTP verbs. The default update method changed from PUT to PATCH, and the `IamPolicy` resources were incorrectly using GET instead of POST for Read operations.

I looked through the proposed changes, and they all seem reasonable.

### CODE REVIEW

File: provider/pkg/gen/schema.go [hunk 1]
```diff
@@ -522,25 +544,48 @@ func (g *packageGenerator) genResource(typeName string, dd discoveryDocumentReso
 				fmt.Printf("Failed to build ID params for resource %s: %v\n", resourceTok, err)
 				return nil
 			}
-			resourceMeta.Read.Endpoint.Template = idPath
-			var idVals []resources.CloudAPIResourceParam
+			// TODO: This logic could be extracted into something more generic. Currently, the value computation
+			//       is always based on the GET method, so this may not be universally correct.
 			for _, k := range codegen.SortedKeys(vals) {
```
Comment (viveklak) [inline provider/pkg/gen/schema.go:549]: I forget why we kept the values as a list in sorted order. Wonder if we should convert it to a map keyed off the name?

File: provider/pkg/resources/resources.go [hunk 2]
```diff
@@ -28,15 +26,75 @@ type CloudAPIMetadata struct {
...
+		// because we flattened the path while building metadata.
+		parts := strings.Split(propValue, "/")
+		propValue = parts[len(parts)-1]
+
+		id = strings.Replace(id, fmt.Sprintf("{%s}", param.Name), propValue, 1)
+	}
+	return id, nil
+}
+
+func evalPropertyValue(values map[string]interface{}, path string) (string, bool) {
```
Comment (viveklak) [inline provider/pkg/resources/resources.go:78]: Would be nice to add a few unit tests for this and other aspects called by the provider?

File: provider/pkg/resources/resources.go [hunk 3]
```diff
@@ -28,15 +26,75 @@ type CloudAPIMetadata struct {
...
+		return v, nil
+	}
+
+	id := e.Template
+	idParams := e.Values
+	for _, param := range idParams {
+		var propValue string
+		if v, has := evalPropertyValue(inputs, param.SdkName); has {
+			propValue = v
+		} else if v, has := evalPropertyValue(outputs, param.SdkName); has {
```
Comment (viveklak) [inline provider/pkg/resources/resources.go:60]: Just trying to understand why/when we would want to fall back to the old value? Is that the actual desired behavior we want?

File: provider/pkg/gen/schema.go (hunk 1 above)
Comment (lblackstone) [reply provider/pkg/gen/schema.go:549]: Yeah, I think that would work also. There's definitely room to simplify this further.

File: provider/pkg/resources/resources.go (hunk 3 above)
Comment (lblackstone) [reply provider/pkg/resources/resources.go:60]: I copied this logic from what we had before. I'm not entirely sure. Perhaps it has to do with properties like `selfLink` that are populated after creation?

File: provider/pkg/resources/resources.go (hunk 2 above)
Comment (lblackstone) [reply provider/pkg/resources/resources.go:78]: Yep, this was a direct copy-paste from a different file, and I didn't include the unit test. I'll see about consolidating the usage.

File: provider/pkg/resources/resources.go (hunk 3 above)
Comment (lblackstone) [reply provider/pkg/resources/resources.go:60]: Confirmed and added a test

File: provider/pkg/gen/schema.go (hunk 1 above)
Comment (lblackstone) [reply provider/pkg/gen/schema.go:549]: This is working for now, so let's follow up on that later

**Review by viveklak — APPROVED** [review]

**Pull request merged**